// @rjsf/core/dist/es/components/widgets/CheckboxesWidget.js
import React from "react";
import PropTypes from "prop-types";

function selectValue(value, selected, all) {
  var at = all.indexOf(value);
  var updated = selected.slice(0, at).concat(value, selected.slice(at)); // As inserting values at predefined index positions doesn't work with empty
  // arrays, we need to reorder the updated selection to match the initial order

  return updated.sort(function (a, b) {
    return all.indexOf(a) > all.indexOf(b);
  });
}

function deselectValue(value, selected) {
  return selected.filter(function (v) {
    return v !== value;
  });
}

function CheckboxesWidget(props) {
  var id = props.id,
      disabled = props.disabled,
      options = props.options,
      value = props.value,
      autofocus = props.autofocus,
      readonly = props.readonly,
      _onChange = props.onChange;
  var enumOptions = options.enumOptions,
      enumDisabled = options.enumDisabled,
      inline = options.inline;
  return (
    /*#__PURE__*/
    React.createElement("div", {
      className: "checkboxes",
      id: id
    }, enumOptions.map(function (option, index) {
      var checked = value.indexOf(option.value) !== -1;
      var itemDisabled = enumDisabled && enumDisabled.indexOf(option.value) != -1;
      var disabledCls = disabled || itemDisabled || readonly ? "disabled" : "";
      var checkbox =
      /*#__PURE__*/
      React.createElement("span", null,
      /*#__PURE__*/
      React.createElement("input", {
        type: "checkbox",
        id: "".concat(id, "_").concat(index),
        checked: checked,
        disabled: disabled || itemDisabled || readonly,
        autoFocus: autofocus && index === 0,
        onChange: function onChange(event) {
          var all = enumOptions.map(function (_ref) {
            var value = _ref.value;
            return value;
          });

          if (event.target.checked) {
            _onChange(selectValue(option.value, value, all));
          } else {
            _onChange(deselectValue(option.value, value));
          }
        }
      }),
      /*#__PURE__*/
      React.createElement("span", null, option.label));
      return inline ?
      /*#__PURE__*/
      React.createElement("label", {
        key: index,
        className: "checkbox-inline ".concat(disabledCls)
      }, checkbox) :
      /*#__PURE__*/
      React.createElement("div", {
        key: index,
        className: "checkbox ".concat(disabledCls)
      },
      /*#__PURE__*/
      React.createElement("label", null, checkbox));
    }))
  );
}

CheckboxesWidget.defaultProps = {
  autofocus: false,
  options: {
    inline: false
  }
};

if (process.env.NODE_ENV !== "production") {
  CheckboxesWidget.propTypes = {
    schema: PropTypes.object.isRequired,
    id: PropTypes.string.isRequired,
    options: PropTypes.shape({
      enumOptions: PropTypes.array,
      inline: PropTypes.bool
    }).isRequired,
    value: PropTypes.any,
    required: PropTypes.bool,
    readonly: PropTypes.bool,
    disabled: PropTypes.bool,
    multiple: PropTypes.bool,
    autofocus: PropTypes.bool,
    onChange: PropTypes.func
  };
}

export default CheckboxesWidget;
